akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -349,20 +349,14 @@ trait FlowOps[+Out, +Mat] {
   import akka.stream.impl.Stages._
   import GraphDSL.Implicits._
 
-  type Repr[+O] <: FlowOps[O, Mat]
+  type Repr[+O] <: FlowOps[O, Mat] {
+    type Repr[+O] = FlowOps.this.Repr[O]
+    type Closed = FlowOps.this.Closed
+  }
 
   // result of closing a Source is RunnableGraph, closing a Flow is Sink
   type Closed
 
-  /*
-   * Repr is actually self-bounded, but that would be a cyclic type declaration that is illegal in Scala.
```

Comment:
This part was wrong: it is not self-bounded, it only needs to produce a type that has the same Repr type constructor. With the fix above scalac will actually verify that implementations are correct.